clients/venice-admin-tool/src/main/java/com/linkedin/venice/KafkaTopicDumper.java

```diff
@@ -234,66 +226,200 @@ public KafkaTopicDumper(
+   * @throws IllegalArgumentException if {@code messageCount} is less than or equal to zero, or if
+   *                                  {@code startOffset} is greater than {@code endOffset}
    */
-  public int fetchAndProcess() {
-    int countdownBeforeStop = maxConsumeAttempts;
-    int currentMessageCount = 0;
-
-    int lastReportedConsumedCount = 0;
-    PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> lastProcessRecord = null;
-    do {
-      Map<PubSubTopicPartition, List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>>> records =
-          consumer.poll(5000); // up to 5 seconds
-      Iterator<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> recordsIterator =
-          Utils.iterateOnMapOfLists(records);
-      while (recordsIterator.hasNext() && currentMessageCount < messageCount) {
-        currentMessageCount++;
-        PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> record = recordsIterator.next();
-        lastProcessRecord = record;
-        processRecord(record);
-      }
+  public int fetchAndProcess(long startOffset, long endOffset, int messageCount) {
+    if (messageCount <= 0) {
+      throw new IllegalArgumentException("Invalid message count: " + messageCount);
+    }
+    if (startOffset > endOffset) {
+      throw new IllegalArgumentException("Invalid offset range: [" + startOffset + ", " + endOffset + ")");
+    }
 
-      if (currentMessageCount - lastReportedConsumedCount > 1000) {
-        LOGGER.info(
-            "Consumed {} messages; last consumed message offset:{}",
-            currentMessageCount,
-            lastProcessRecord.getOffset());
-        lastReportedConsumedCount = currentMessageCount;
+    int remainingAttempts = maxConsumeAttempts;
+    int processedMessageCount = 0;
+    int lastReportedCount = 0;
+
+    consumer.subscribe(topicPartition, startOffset - 1);
+
+    try {
+      PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> lastProcessedRecord = null;
+      while (remainingAttempts > 0 && processedMessageCount < messageCount) {
+        // Poll for records
+        Map<PubSubTopicPartition, List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>>> records =
+            consumer.poll(5000); // Poll for up to 5 seconds
+        Iterator<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> recordIterator =
+            Utils.iterateOnMapOfLists(records);
+        boolean hasProcessedRecords = false;
+        while (recordIterator.hasNext() && processedMessageCount < messageCount) {
+          PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> record = recordIterator.next();
+          processedMessageCount++;
+          lastProcessedRecord = record;
+          hasProcessedRecords = true;
+          processRecord(record);
+          // Exit early if endOffset is reached
+          if (record.getOffset() >= endOffset) {
+            LOGGER.info("Reached endOffset. Total messages processed: {}", processedMessageCount);
+            return processedMessageCount;
```

Comment:
Can we just break out of the loop here?

Reply:
We can but since this is nested loop we'll have to repeat the condition in both inner and outer loop conditions

Reply:
Oh, didn't realize it's a nested loop